File: 0301-remove-invalid-parentheses/0301-remove-invalid-parentheses.cpp
class Solution {
public:
    bool valid(string s)
    {
        int balance = 0;
        for(auto& c : s)
        {
            if(c == '(') balance++;
            else if(c == ')') balance--;
            
            if(balance < 0) return false;
        }
        
        return balance == 0;
    }
    
    void recur(int start, int l, int r, string s, vector<string>& ans)
    {
        if(l == 0 && r == 0)
        {
            if(valid(s))
            {
                ans.push_back(s);
            }
            
            return;
        }
        
        
        for(int i = start; i < s.length(); i++)
        {
            if(i != start && s[i] == s[i - 1]) continue;
            
            if(s[i] == '(' || s[i] == ')')
            {
                string news = s.substr(0, i) + s.substr(i + 1, s.length() - (i + 1));
            
                if(s[i] == '(' && l > 0)
                    recur(i, l - 1, r, news, ans);
                else if(s[i] == ')' && r > 0)
                    recur(i, l, r - 1, news, ans);
            }
        }
    }
    
    vector<string> removeInvalidParentheses(string s)
    {
        int l = 0, r = 0;
        
        for(auto& c : s)
        {
            if(c == '(') l++;
            
            if(c == ')')
            {
                if(l == 0) r++;
                else l--;
            }
        }
        
        vector<string> ans;
        
        recur(0, l, r, s, ans);
        return ans;
    }
};

File: 0301-remove-invalid-parentheses/0301-remove-invalid-parentheses.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        l, r = 0, 0
        
        for c in s:
            if c == '(':
                l += 1
            elif c == ')':
                if l > 0:
                    l -= 1
                else:
                    r += 1
        
        ans = []
        
        
        def valid(s):
            balance = 0
            
            for c in s:
                if c == '(':
                    balance += 1
                elif c == ')':
                    balance -= 1
                    
                if balance < 0:
                    return False
            
            return balance == 0
        
        def dfs(start, s, l, r):
            if l == 0 and r == 0:
                if valid(s):
                    ans.append(s[:])
                return
            
            for i in range(start, len(s)):
                if i != start and s[i] == s[i - 1]:
                    continue
                
                if s[i] == '(' or s[i] == ')':
                    newS = s[:i] + s[i + 1:]

                    if s[i] == '(' and l > 0:
                        dfs(i, newS, l - 1, r)
                    elif s[i] == ')' and r > 0:
                        dfs(i, newS, l, r - 1)
        
        
        dfs(0, s, l, r)
        
        return ans
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            
            